daml-lf/spec/daml-lf-1.rst

```diff
@@ -3151,69 +3134,148 @@ in which keys were inserted into the map (insertion-order).
 
+  Formally the builtin function ``GENMAP_DELETE`` semantics is defined
+  by the following rules. ::
+
+      𝕆('EQUAL' @σ vᵢ v) = Err t  for some i ∈ 1, …, n
+    —————————————————————————————————————————————————————————————————————— EvGenMapDeleteErr
+      𝕆('GENMAP_DELETE' @σ @τ 〚v₁ ↦ w₁; … ; vₙ ↦ wₙ〛 v) = Err t
+
+      𝕆('EQUAL' @σ vᵢ v) = Ok 'True'  for some i ∈ 1, …, n
+    —————————————————————————————————————————————————————————————————————— EvGenMapDeletePresent
+      𝕆('GENMAP_DELETE' @σ @τ 〚v₁ ↦ w₁; … ; vₙ ↦ wₙ〛 v) =
+        Ok' 〚v₁ ↦ w₁; … ; vᵢ₋₁ ↦ wᵢ₋₁; vᵢ₊₁ ↦ wᵢ₊₁; … ; vₙ ↦ wₙ〛
+
+      𝕆('EQUAL' @σ vᵢ v) = Ok 'False'  for all i ∈ 1, …, n
+    —————————————————————————————————————————————————————————————————————— EvGenMapDeleteAbsent
+      𝕆('GENMAP_DELETE' @σ @τ 〚v₁ ↦ w₁; … ; vₙ ↦ wₙ〛 v) =
+        'Ok' 〚v₁ ↦ w₁; … ; vₙ ↦ wₙ〛
+
 * ``GENMAP_KEYS : ∀ α. ∀ β.  'GenMap' α β → 'List' α``
 
-  Get the list of keys in the map. The keys are returned by insertion
-  order, so if you insert key ``x`` before key ``y``, then ``x`` will
-  appear before ``y`` in the list.
+  Get the list of keys in the map. The keys are returned in the order
+  they appear in the map.
 
   [*Available in versions >= 1.dev*]
 
+  Formally the builtin function ``GENMAP_KEYS`` semantics is defined
+  by the following rules. ::
+
+    —————————————————————————————————————————————————————————————————————— EvGenMapKeysEmpty
+      𝕆('GENMAP_KEYS' @σ @τ 〚〛) = 'Ok' (Nil @σ)
+
+      𝕆('GENMAP_KEYS' @σ @τ 〚v₁ ↦ w₁; … ; vₙ ↦ wₙ〛) = 'Ok' vₗ
+    —————————————————————————————————————————————————————————————————————— EvGenMapKeysNonEmpty
+      𝕆('GENMAP_KEYS' @σ @τ 〚v₀ ↦ w₀; v₁ ↦ w₁; … ; vₙ ↦ wₙ〛) =
+        'Ok' (Cons @σ v₀ vₗ)
+
 * ``GENMAP_VALUES : ∀ α. ∀ β.  'GenMap' α β → 'List' β``
 
   Get the list of values in the map. The values are returned in the
-  same order as ``GENMAP_KEYS``, so the ith element of ``GENMAP_KEYS``
-  maps to the ith element of ``GENMAP_VALUES``.
+  order they appear in the map (i.e. sorted by key).
 
   [*Available in versions >= 1.dev*]
 
+  Formally the builtin function ``GENMAP_VALUES`` semantics is defined
+  by the following rules. ::
+
+    —————————————————————————————————————————————————————————————————————— EvGenMapValuesEmpty
+      𝕆('GENMAP_VALUES' @σ @τ 〚〛) = 'Ok' (Nil @τ)
+
+      𝕆('GENMAP_VALUES' @σ @τ 〚v₁ ↦ w₁; … ; vₙ ↦ wₙ〛) = 'Ok' wₗ
+    —————————————————————————————————————————————————————————————————————— EvGenMapValuesNonEmpty
+      𝕆('GENMAP_KEYS' @σ @τ 〚v₀ ↦ w₀; v₁ ↦ w₁; … ; vₙ ↦ wₙ〛) =
+        'Ok' (Cons @τ w₀ wₗ)
+  
 * ``GENMAP_SIZE : ∀ α. ∀ β.  'GenMap' α β → 'Int64'``
 
   Return the number of elements in the map.
 
   [*Available in versions >= 1.dev*]
 
-
-**Validity of Keys:** A key is valid if and only if it is equivalent
-to itself according to the relation ``~ᵥ`` defined below. Attempts to
-use an invalid key in the operations above always result in a runtime
-error.
```

Comment:
You can simply reinstate this paragraph if you want to leave the possibility for hashtable-based implementation open (but use ``EQUAL`` instead of ``~ᵥ``).

Reply:
done